src/modules/brightness.rs

```diff
@@ -0,0 +1,351 @@
+use crate::channels::{AsyncSenderExt, BroadcastReceiverExt};
+use crate::clients::brightness::{self, brightness, default_resource_name, max_brightness};
+use crate::config::{CommonConfig, LayoutConfig, TruncateMode};
+use crate::modules::{Module, ModuleInfo, ModuleParts, WidgetContext};
+use crate::{module_impl, spawn};
+use color_eyre::Result;
+use gtk::{Button, Label};
+use gtk::{EventControllerScroll, EventControllerScrollFlags, prelude::*};
+use serde::Deserialize;
+use tokio::sync::mpsc;
+use tokio::time::sleep;
+
+#[derive(Debug, Clone, Deserialize)]
+#[cfg_attr(feature = "extras", derive(schemars::JsonSchema))]
+pub enum BrightnessDataSource {
+    /// using the keyboard dbus. Note: this only works for keyboards, not for screen brightness.
+    Keyboard,
+    /// using the login1 dbus and fs for reading. This works for keyboard and screen brightness, but needs the filesystem for reading the data and dbus for adjusting.
+    Login1Fs {
+        /// The subsystem to read the data from, e.g. `backlight` or `leds`. Subsystem refers to the directory within `/sys/class/`.
+        ///
+        /// **Default**: `backlight`
+        subsystem: String,
+
+        /// The name of the resource, see `/sys/class/<subsystem>` for available resources. If empty, ironbar will try to resolve it via a hardcoded list of common resources
+        ///
+        /// **Default**: `None`
+        name: Option<String>,
+    },
+}
+
+#[derive(Debug, Clone, Deserialize)]
+#[cfg_attr(feature = "extras", derive(schemars::JsonSchema))]
+#[serde(default)]
+pub struct BrightnessModule {
+    /// The format string to use for the widget button label.
+    /// For available tokens, see [below](#formatting-tokens).
+    ///
+    /// **Default**: `{icon} {percentage}%`
+    format: String,
+
+    /// Brightness state icons.
+    ///
+    /// See [icons](#icons).
+    icons: Icons,
+
+    /// Where to get the brightness data from
+    ///
+    /// See [BrightnessDataSource].
+    datasource: BrightnessDataSource,
```

Comment:
`datasource` feels a bit technical for a config option. I'd prefer `mode` for consistency with existing widgets.

```suggestion
    mode: BrightnessDataSource,
```